_posts/2020-04-23-spring-boot-testcontainers.md

```diff
@@ -0,0 +1,215 @@
+other databases.
+At this moment we could take a not clean way and create two sets of SQL scripts for migration.
+It could look like this.
+
+![Two sets of SQL scripts.](/assets/img/posts/testcontainers/two_sets_of_sql_scripts.png)
+
+Now we :
+ * have to configure Spring Boot profiles for different folders with scripts,
+ * have to maintain two sets of scripts,
+ * are not able to test scripts from the folder `postgresql` at build time.
+ 
+If we want to write a new script with some features, that are not supported by `H2`,
+we have to write two scripts, one for `H2` and one for `PostgreSQL`. Also,
+we have to find a way to achieve the same results with both
+scripts.
+
+If we test the migration with `H2` database, and our test is green, we don't know anything about the script `V1_2__change_columnt_type.sql` from
+folder `postgresql`.
+
+Again, even when we don't have different sets of scripts for the migration, the green test doesn't mean, that they can run
+on the `PostgresSQL` database.
+
+## Using Production-Like Environment for Migration Tests 
+There is another approach for testing database migration. First, we can test migration with `H2` database at build time.
+After that, we can deploy our application in a production-like environment and let the migration run on this environment with 
+the database like in production, for example, `PostgreSQL`.
+
+This approach has a couple of drawbacks:
+
+* Bugs are discovered too late,
+* it is hard to find errors.
+
+Let's imagine, we test the migration with `H2` database during building the application, and the tests are green.
+The next step is delivering and deploying the application to a test environment. It takes time. If the migration on the test environment fails,
+we know it too late, maybe several minutes later. It slows down the development cycle.
+
+Also, this situation is very confusing for developers, because we can't debug this error with our unit test.
+Our unit test with `H2` was green, and the error occurred on the test environment.
+
+## Solution with Testcontainers
+With [Tescontainers](https://www.testcontainers.org/) we can test the database migration with [Docker](https://www.docker.com/) from our code.
+`Tescontainers` is a Java library, that provides unit testing with Docker.
+Of course, we have to install Docker wherever we want to build the application. After that we can create an initialization code for testing:
+
+````java
+@ContextConfiguration(initializers 
+                               = AbstractIntegrationTest.Initializer.class)
+public class AbstractIntegrationTest {
+
+    static class Initializer
+            implements ApplicationContextInitializer
+                        <ConfigurableApplicationContext> {
+
+        static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>();
+
+        public static Map<String, String> getProperties() {
```

Comment:
This method is hard to explain to a reader with no background because it mixes responsibilities. It starts the postgres container and maps the properties into Spring properties. This should be two methods, which should be easier to explain in the text :). 

Also see my comment below.

Reply:
I refactored the code and change the explanation.